### ISSUE TITLE
[PROPOSAL]: support not implementation of delta lake or some other physical manifestation, but V2 data source of Spark

### ISSUE THREAD

**Issue by MironAtHome** [microsoft/hyperspace#513]
## Problem Statement
A clear and concise description of what the problem is e.g., *I have this scenario and Hyperspace does not work [...]*
There seems to be a gaping hole in the architectural approach in designing spark physical data access. It is somewhat natural and totally inline with how developer would plunge into delivering a working solution. Unawares though of a greater design picture.
I believe it is time to take a step back and look at a Spark not as a wonder that has capabilities, but as a design, a blueprint, or a total sum of its blueprints, which holds the key to its purpose and approach to implementing its features.
One of the key blueprints underpinning entire Spark physical data access path ( and indexing is a nothing, but the point of the razor, not only accessing physical data, but optimizing this access to the finite degree, is the V2 Data Source interface, or collection interfaces defining physical layer of data access of Spark.
It was lurking under the copious amount of code all the way since version 2.3.1. I believe it is time to dust away the obfuscating details an implement this V2 interface, albeit limited.

## Background and Motivation
An introduction of the necessary background and the problem being solved by the proposed change.

## Proposed Solution
A clear and concise sketch of the solution you want to propose. Please keep this short. There is a section below for writing a detailed solution.

## Known/Potential Compatibility Issues
A discussion of the change with regard to the compatibility guidelines.

## Design
A description of the proposed design/algorithm. This should include a discussion of how the work fits into Hyperspace's roadmap.

## Implementation
A description of the steps in the implementation, who will do them, and when.

## Performance Implications (if applicable)
A discussion of impact on performance and any corner cases that the author is aware of. If there is a negative impact on performance, please make sure to capture an issue in the next section. This section may be omitted if there are none.

**Open issues (if applicable)**
A discussion of issues relating to this proposal for which the author does not know the solution. If you have already opened the corresponding issues, please link to them here. This section may be omitted if there are none.

  - This is the first issue (issue-link)
  - This is the second issue (issue-link)
  - ...

**Additional context (if applicable)**

**Comment by clee704** [comment]
I agree that it could be more beneficial for the community in general in the long run if Spark supports index as a concept in Data Source V2 API. Actually, there is already work going on ([SPARK-36525](https://issues.apache.org/jira/browse/SPARK-36525)). If you want to contribute to that direction, then it might be better to work on Spark directly.

I think one of the values of Hyperspace is being or trying to be data format agnostic. For example, covering indexes work as long as the data source supports creating from an existing dataset a new dataset with fewer columns and laid out in certain ways (i.e. bucketing) for efficient scanning by selected columns. Data skipping indexes work as long as the data source stores data in multiple objects (i.e. files) and can compute aggregations grouped by object. So we support not only Parquet, Delta Lake, Iceberg, but also CSV and other data sources as long as they are capable of certain things.

Looking at the current code and how it's evolving, it seems the index API in Spark doesn't allow Hyperspace-like indexing subsystems that support more than one data source to be plugged in. If we want to build Hyperspace around the Data Source V2 API, then we should propose a suitable change to the API so that for example we can hook into the index API.
